### PR TITLE
Improve Shadertoy Utility

### PR DESCRIPTION
**Pull request by Vipitis**
picking up from https://github.com/pygfx/wgpu-py/issues/360

Several steps to this PR (will be updated)
- [x] add offscreen support
- [x] add snapshot method
- [x] ~~validate shadercode (with naga?) before creating shader module~~ (addressed by https://github.com/gfx-rs/wgpu-native/pull/313)
- [x] add tests for new features
- [x] fix style
- [ ] rebase

I welcome any feedback to my additions. I don't want to break any existing functionality.

### CODE REVIEW

File: wgpu/utils/shadertoy.py [hunk 1]
```diff
@@ -1,9 +1,14 @@
 import time
 import ctypes
+import numpy as np
+from PIL import Image
```
Comment (almarklein) [inline wgpu/utils/shadertoy.py:4]: This gets a bit tricky in terms of dependencies. I'd prefer to make `snapshot()` return a memoryview object, that can be easily converted to a numpy array (can mention in the docs that ppl can use `np.asarray()`.

Comment (Vipitis) [reply wgpu/utils/shadertoy.py:4]: definitely, I haven't even looked up what dependencies are included, and looks like numpy is just in the dev requirements. 
Will shorten the method and the copy over the docstring

**Comment by Vipitis** [comment]
short update, I had encountered more panics and found a ways to catch them by doing more translations.
And now that glsl errors get handled properly, it really shouldn't be here anymore.

Would it be okay to take all of that validation out (I will keep it in own project instead) and just add the offscreen and snapshot features with this PR for now?

File: wgpu/utils/shadertoy.py [hunk 2]
```diff
@@ -463,8 +474,67 @@ def _draw_frame(self):
...
         self._canvas.request_draw(self._draw_frame)
-        run()
+        if self._offscreen:
+            run_offscreen()
+        else:
+            run()
+
+    def snapshot(self, time_float: float = 0.0, mouse_pos: tuple = (0, 0, 0, 0)):
+        """
+        Returns a PIL Image of the specified time. (Only available when ``offscreen=True``)
```
Comment (almarklein) [inline wgpu/utils/shadertoy.py:484]:
```suggestion
        Returns an image of the specified time. (Only available when ``offscreen=True``)
```

**Comment by almarklein** [comment]
> Would it be okay to take all of that validation out (I will keep it in own project instead) and just add the offscreen and snapshot features with this PR for now?

Yes, that sounds good! The glsl validation are present in the latest wgpu-native release, so they will be in wgpu-py in the next update.

**Comment by almarklein** [comment]
Thanks!

**Pull request merged**